diff -Naur /tmp/dwm2/src/dwm.c src/dwm.c
--- /tmp/dwm2/src/dwm.c	2012-07-11 20:25:13.733157465 +0200
+++ src/dwm.c	2012-07-11 19:56:57.659697028 +0200
@@ -1518,11 +1518,34 @@
 void
 run(void) {
 	XEvent ev;
+
+    fd_set in_fds;
+    struct timeval tv;
+    tv.tv_usec = 0;
+
+    int X11_fd = ConnectionNumber(dpy);
+    time_t now, next_update = time(NULL);
 	/* main event loop */
 	XSync(dpy, False);
-	while(running && !XNextEvent(dpy, &ev))
-		if(handler[ev.type])
-			handler[ev.type](&ev); /* call handler */
+	while(running) {
+        now = time(NULL);
+        if (next_update <= now) {
+            updatestatus();
+            next_update = now + status_nextupdatein();
+        }
+
+        FD_ZERO(&in_fds);
+        FD_SET(X11_fd, &in_fds);
+        tv.tv_sec = next_update - now;
+
+        if (XPending(dpy) || select(X11_fd +1, &in_fds, 0, 0, &tv)) {
+            while(XPending(dpy)) {
+                XNextEvent(dpy, &ev);
+                if(handler[ev.type])
+                    handler[ev.type](&ev); /* call handler */
+            }
+        }
+	}
 }
 
 void
@@ -2125,7 +2148,8 @@
 void
 updatestatus(void) {
 	if(!gettextprop(root, XA_WM_NAME, stext, sizeof(stext)))
-		strcpy(stext, "dwm-"VERSION);
+		strcpy(stext, "");
+    status_complete(stext, sizeof(stext) - 1);
 	drawbar(selmon);
 }
 
diff -Naur /tmp/dwm2/src/dwm.c.orig src/dwm.c.orig
--- /tmp/dwm2/src/dwm.c.orig	2012-07-11 20:25:13.726490801 +0200
+++ src/dwm.c.orig	2012-07-11 19:56:57.656363695 +0200
@@ -1518,11 +1518,34 @@
 void
 run(void) {
 	XEvent ev;
+
+    fd_set in_fds;
+    struct timeval tv;
+    tv.tv_usec = 0;
+
+    int X11_fd = ConnectionNumber(dpy);
+    time_t now, next_update = time(NULL);
 	/* main event loop */
 	XSync(dpy, False);
-	while(running && !XNextEvent(dpy, &ev))
-		if(handler[ev.type])
-			handler[ev.type](&ev); /* call handler */
+	while(running) {
+        now = time(NULL);
+        if (next_update <= now) {
+            updatestatus();
+            next_update = now + status_nextupdatein();
+        }
+
+        FD_ZERO(&in_fds);
+        FD_SET(X11_fd, &in_fds);
+        tv.tv_sec = next_update - now;
+
+        if (XPending(dpy) || select(X11_fd +1, &in_fds, 0, 0, &tv)) {
+            while(XPending(dpy)) {
+                XNextEvent(dpy, &ev);
+                if(handler[ev.type])
+                    handler[ev.type](&ev); /* call handler */
+            }
+        }
+	}
 }
 
 void
@@ -2123,7 +2146,8 @@
 void
 updatestatus(void) {
 	if(!gettextprop(root, XA_WM_NAME, stext, sizeof(stext)))
-		strcpy(stext, "dwm-"VERSION);
+		strcpy(stext, "");
+    status_complete(stext, sizeof(stext) - 1);
 	drawbar(selmon);
 }
 
diff -Naur /tmp/dwm2/src/status.c src/status.c
--- /tmp/dwm2/src/status.c	1970-01-01 01:00:00.000000000 +0100
+++ src/status.c	2012-07-11 20:37:40.089880065 +0200
@@ -0,0 +1,171 @@
+/* Functions to keep the dwm status bar updated within dwm itself.
+   status_complete is called when it is time to update the status bar.
+
+   To configure: edit status_complete to your liking. Then configure
+   status_nextupdatein to set how often the status bar should update.
+
+   Compile with -DASTEST to enable testing. main() is then included
+   and will create an executable that prints a couple of 
+   dummy status lines, which is useful for testing.*/
+
+#include <dirent.h>
+#include <signal.h>
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h> 
+static int
+strlcat(char *dest, const char *src, int maxlength)
+{
+    strncat(dest, src, maxlength - strlen(dest));
+    return strlen(dest);
+}
+
+static int
+readint(const char *filename) {
+    int result = -1;
+    FILE* fp = fopen(filename, "r");
+
+    if (fp) {
+        fscanf(fp, "%d", &result);
+        fclose(fp);
+    }
+    return result;
+}
+
+static void
+status_appendtime(char *stext, const char* format, const char *separator, int maxlen) {
+    time_t now;
+    struct tm *tminfo;
+    char timestr[100];
+
+    now = time(NULL);
+    tminfo = localtime(&now);
+
+    if (strftime(timestr, sizeof(timestr), format, tminfo)) {
+        if (stext[0] != '\0')
+            strlcat(stext, separator, maxlen);
+        strlcat(stext, timestr, maxlen);
+    }
+}
+
+static void
+status_appendbattery(char *stext, const char *separator, int maxlen)
+{
+    char battxt[6];
+    battxt[0] = '\0';
+
+    FILE *fp = fopen("/sys/class/power_supply/BAT0/status", "r");
+    if (fp) {
+        int status = fgetc(fp);
+        fclose(fp);
+
+        int now = readint("/sys/class/power_supply/BAT0/energy_now");
+        int full = readint("/sys/class/power_supply/BAT0/energy_full");
+
+        if (now > 0 && full > 0) {
+            int charge = now/(full / 100);
+            char charge_char = 0x00;
+            if (status == 'C')
+                charge_char = '+';
+            else if (status == 'D')
+                charge_char = '-';
+
+            if (charge_char) {
+                sprintf(battxt, "%c%d%%", charge_char, charge);
+            }
+            else {
+                sprintf(battxt, "%d%%", charge);
+            }
+        }
+    }
+
+    if (battxt[0] != '\0') {
+        if (stext[0] != '\0')
+            strlcat(stext, separator, maxlen);
+        strlcat(stext, battxt, maxlen);
+    }
+}
+
+/* netcfg connected network */
+static void
+status_appendnetwork(char *stext, const char *separator, int maxlen)
+{
+    char netsstr[256];
+    char *filename;
+    DIR *dir=opendir("/run/network/profiles");
+    struct dirent *ent;
+
+    netsstr[0] = '\0';
+    if (dir) {
+        while((ent=readdir(dir)) != NULL) {
+            filename = ent->d_name;
+            if (filename[0] != '.') {
+                strlcat(netsstr, ent->d_name, sizeof(netsstr) - 1);
+                strlcat(netsstr, " ", sizeof(netsstr) - 1);
+            }
+        }
+        closedir(dir);
+
+        if (netsstr[0] != '\0') {
+            /* Remove space after last network */
+            netsstr[strlen(netsstr)-1] = '\0';
+            if (stext[0] != '\0')
+                strlcat(stext, separator, maxlen);
+            strlcat(stext, netsstr, maxlen);
+        }
+    }
+}
+
+/* Edit status_complete to configure what your status line should look like.
+   stext contains what dwm traditionally puts on the status line, i.e. the X
+   root window name. You should modify stext, which will in the end be printed
+   on the status bar. If you for example append text to stext, it will show
+   after the traditional status bar. */
+static void
+status_complete(char *stext, int maxlen)
+{
+    const char* separator = " >> ";
+    status_appendnetwork(stext, separator, maxlen);
+    status_appendbattery(stext, separator, maxlen);
+    status_appendtime(stext, "%R", separator, maxlen);
+}
+
+/* When should the status bar be updated next? This is called after an update,
+   to allow updates at e.g. top of the minute.*/
+static int
+status_nextupdatein()
+{
+	/* Update at the top of the minute */
+	time_t now = time(NULL);
+	struct tm *tminfo = localtime(&now);
+
+	return 60 - tminfo->tm_sec;
+}
+
+#ifdef ASTEST
+
+#include <stdio.h>
+
+int main()
+{
+    char line[100];
+
+    /* With regular status text */
+    strcpy(line, "status text");
+    status_complete(line, sizeof(line) - 1);
+    printf("%s\n", line);
+
+    /* With no status text */
+    strcpy(line, "");
+    status_complete(line, sizeof(line) - 1);
+    printf("%s\n", line);
+
+    /* With limited length, i.e. truncated */
+    strcpy(line, "");
+    status_complete(line, 6);
+    printf("%s\n", line);
+
+    return 0;
+}
+#endif
